### PR TITLE
 stdenv: move --enable-deterministic-archives flag into GNU wrapper 

### LINKED ISSUE #138013: clangStdenv does not strip on Linux
The Linux stdenv sets

https://github.com/NixOS/nixpkgs/blob/722c8c9529baadd73fb0e5cf1ec896ca2a90b8f3/pkgs/stdenv/linux/default.nix#L379-L381

But that flag is specific to GNU strip, and LLVM's strip will fail if passed it.  Conveniently, we silence all errors from stripping, so this has gone unnoticed until now:

https://github.com/NixOS/nixpkgs/blob/722c8c9529baadd73fb0e5cf1ec896ca2a90b8f3/pkgs/build-support/setup-hooks/strip.sh#L54

I noticed this while trying to fix `pkgsLLVM.python3`.  One of the reasons it doesn't work is that it has `openssl.dev` in `disallowedRequisites`, and strip should be removing that reference but isn't. 

What I'm not sure about here is the best way to fix this.  There's nothing platform-specific about this option, so it shouldn't live in the Linux stdenv.  (Other platforms like NetBSD use GNU binutils by default.)  Ideally it would be a conditional in the strip setup hook, but we don't really have a bash equivalent to `stdenv.cc.isGNU`.  So where (in Nix) should we check this and set the variable?

cc @sternenseemann

### PR DESCRIPTION
**Pull request by sternenseemann**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/stdenv/linux/default.nix [hunk 1]
```diff
@@ -375,12 +375,7 @@ in
       targetPlatform = localSystem;
       inherit config;
 
-      preHook = ''
-        # Make "strip" produce deterministic output, by setting
-        # timestamps etc. to a fixed value.
-        commonStripFlags="--enable-deterministic-archives"
```
Comment (Ericson2314) [inline pkgs/stdenv/linux/default.nix:381]: Based on a grep, you can get rid of `commonStripFlags` altogether. (It's defined here and used in pkgs/build-support/setup-hooks/strip.sh`.)

File: pkgs/development/interpreters/python/cpython/default.nix [hunk 2]
```diff
@@ -286,7 +286,9 @@ in with passthru; stdenv.mkDerivation {
   CPPFLAGS = concatStringsSep " " (map (p: "-I${getDev p}/include") buildInputs);
   LDFLAGS = concatStringsSep " " (map (p: "-L${getLib p}/lib") buildInputs);
   LIBS = "${optionalString (!stdenv.isDarwin) "-lcrypt"} ${optionalString (ncurses != null) "-lncurses"}";
-  NIX_LDFLAGS = optionalString (stdenv.isLinux && !stdenv.hostPlatform.isMusl) "-lgcc_s" + optionalString stdenv.hostPlatform.isMusl "-lgcc_eh";
+  NIX_LDFLAGS =
+    optionalString (stdenv.isLinux && !stdenv.hostPlatform.isMusl && !(stdenv.hostPlatform.useLLVM or false)) "-lgcc_s"
+    + optionalString (stdenv.hostPlatform.isMusl && !(stdenv.hostPlatform.useLLVM or false)) "-lgcc_eh";
```
Comment (alyssais) [inline pkgs/development/interpreters/python/cpython/default.nix:291]: I had

```nix
if stdenv.isLinux && stdenv.cc.isGNU then
  if stdenv.hostPlatform.isMusl then
    "-lgcc_eh"
  else
    "-lgcc_s"
else
  ""
```

I think it's more readable that way.

Comment (sternenseemann) [reply pkgs/development/interpreters/python/cpython/default.nix:291]: fair enough, I just hacked this in. I think `stdenv.isLinux && stdenv.cc.isGNU` can be simplified to just `stdenv.cc.isGNU` in practice? Hard to tell whether there is actual merit to the `isLinux` thing or it was just a case of not darwin.

**Comment by Ericson2314** [comment]
> TODO can we get rid of this NIX_LDFLAGS hack altogether?

Where is that?

> We can at the moment, but that whole system is being reworked in #132343, so I'd rather not touch anything while that's ongoing.

I am not sure when I'll finish that (as much as I like it!) so without knowing the details, if there is some hack to get rid of now easily, by all means do so.

**Comment by sternenseemann** [comment]
> Where is that?

76c9a926e06afd07b2a263504f60e64b0675fea7 which Alyssa commented on (a fact that easily gets lost in the github UI…)

**Comment by Ericson2314** [comment]
Oh I see. Yeah fine for now for sure.

File: pkgs/development/interpreters/python/cpython/default.nix (hunk 2 above)
Comment (alyssais) [reply pkgs/development/interpreters/python/cpython/default.nix:291]: I doubt it's necessary with e.g. gcc on Darwin.  So I think we should first check if cc is gcc, and then say if it's Musl do gcc_eh, and if it's Glibc do gcc_s.

Comment (sternenseemann) [reply pkgs/development/interpreters/python/cpython/default.nix]: Done.

File: pkgs/stdenv/linux/default.nix (hunk 1 above)
Comment (sternenseemann) [reply pkgs/stdenv/linux/default.nix:381]: Done.

**Comment by sternenseemann** [comment]
I think this is good enough for now. Will need some additional testing though, as it is a full stdenv rebuild on all platforms now.

**Review by alyssais — APPROVED** [review]

**Review by Ericson2314 — APPROVED** [review]

**Pull request merged**